### PR TITLE
Make metrics test utils a bit more flexible

### PR DESCRIPTION
**Pull request by maximsch2**
I'm testing a metric that doesn't make much sense with compute_on_step=True (another PR incoming with it a bit later) and doing a few small changes here:
1) First lets me `assert not compute_on_step` in the metric by only passing it to the metric in tests if asked
2) Reduces the number of calls to sk_metric when we don't need them.

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: tests/metrics/utils.py [hunk 1]
```diff
@@ -104,14 +108,14 @@ def _class_test(
             if rank == 0:
```
Comment (rohitgr7) [inline tests/metrics/utils.py:108]:
```suggestion
            if rank == 0 and check_dist_sync_on_step:
                ddp_preds = ...
                ddp_target = ...
                sk_batch_result = ..
                _assert_...
```
can be done like this since there is no else condition here?

File: tests/metrics/utils.py [hunk 2]
```diff
@@ -104,14 +108,14 @@ def _class_test(
...
                 ddp_target = torch.cat([target[i + r] for r in range(worldsize)])
-                sk_batch_result = sk_metric(ddp_preds, ddp_target)
                 # assert for dist_sync_on_step
                 if check_dist_sync_on_step:
+                    sk_batch_result = sk_metric(ddp_preds, ddp_target)
                     _assert_allclose(batch_result, sk_batch_result, atol=atol)
         else:
-            sk_batch_result = sk_metric(preds[i], target[i])
             # assert for batch
             if check_batch:
```
Comment (carmocca) [inline tests/metrics/utils.py:117]: merge into elif?

**Comment by maximsch2** [comment]
How do I find things that are losing coverage from this bot?

**Comment by SkafteNicki** [comment]
@maximsch2 do not worry about the coverage, we been having trouble with it lately

**Review by ananyahjha93 — APPROVED** [review]

**Comment by Borda** [comment]
Hi @maximsch2, we are very happy that you decided to improve PL and opened this PR.
Never the less we have moved the metric package outside PT, to a separate repository - [torchmetrics](https://github.com/PyTorchLightning/metrics), and unfortunately, there is not GH option to transfer your opened PR there to finish the job there. So I would like to kindly ask you if you are able to finish this PR by end of this week and then we can transfer your merged commit to the new repo or please open the very same PR in the new repository... 🐰 

Anyway thank you very much for being with us ⚡

**Review by justusschock — APPROVED** [review]

**Comment by maximsch2** [comment]
Is it possible to trigger tests to re-run? The circleci failure seems transient.

**Comment by SkafteNicki** [comment]
@maximsch2 i added an entry to the changelog regarding your PR, which retriggered tests

**Review by s-rog — APPROVED** [review]

**Comment by Borda** [comment]
@maximsch2 as we already started the deprecation here in PL it would be even more difficult to finish adding a new metric here, may I kindly ask you to reopen the very same PR in `torchmetrics` (the reason is to preserve you as the contribution author) and I will help you to finish it there :]

So pls, open PR in TM with the very same content and refer to this PL PR 🐰

**Comment by maximsch2** [comment]
Sounds good, let me move this over.
